File: examples/e2e-consent-demo-py/sp_shared/serve.py
# Boots one SP app from its provisioned identity. Both sp_airline/server.py
# and sp_hotel/server.py are thin wrappers around this. Python port of
# sp-shared/serve.ts.

from __future__ import annotations

import os

from helixid_config import Env, SpDefinition, env as default_env
from sp_shared.app import create_sp_app
from sp_shared.audit import create_audit_emitter
from sp_shared.identity import STATUS_LIST_LENGTH, load_sp_identity, state_path
from sp_shared.status_list import build_status_list_credential, create_status_list
from sp_shared.store import SpStore

HERE = os.path.dirname(os.path.abspath(__file__))


def serve_sp(definition: SpDefinition, env: Env = default_env) -> None:
    identity = load_sp_identity(env.wallets_dir, definition.id)
    store = SpStore.open(
        state_path(env.wallets_dir, definition.id),
        build_status_list_credential("1", create_status_list(STATUS_LIST_LENGTH), identity.did, identity.baseUrl),
    )

    def on_audit_error(message: str) -> None:
        print(f"[{definition.id}] {message}", flush=True)

    app, _counters = create_sp_app(
        definition=definition,
        issuer_did=identity.did,
        issuer_private_key_hex=identity.privateKeyHex,
        issuer_public_key_hex=identity.publicKeyHex,
        base_url=identity.baseUrl,
        helix_api_url=env.helix_api_url,
        store=store,
        # Scope resolution happens inside this same process. The public
        # did:web base URL can be localhost, but the server should call its
        # own loopback.
        mcp_server_url=f"http://127.0.0.1:{definition.port}/api/mcp",
        # @helixid/widget has no Python port -- its pre-built browser bundle
        # is served as-is. docker/python.Dockerfile builds it from the public
        # helix-sdk-js repo into ../widget-dist; override WIDGET_DIST_PATH to
        # point at your own checkout when running outside Docker.
        widget_dist_path=os.environ.get(
            "WIDGET_DIST_PATH", os.path.join(HERE, "..", "widget-dist")
        ),
        audit=create_audit_emitter(
            env.helix_api_url,
            env.admin_api_key,
            identity.did,
            definition.display_name,
            on_error=on_audit_error,
        ),
    )

    print(f"[{definition.id}] {definition.display_name} listening on :{definition.port}", flush=True)
    print(f"[{definition.id}] did:web  {identity.did}", flush=True)
    print(f"[{definition.id}] status   {identity.statusListUrl}", flush=True)
    print(f"[{definition.id}] consent  {identity.baseUrl}/consent", flush=True)
    app.run(host="0.0.0.0", port=definition.port)

File: examples/e2e-consent-demo/sp-shared/serve.ts
// Boots one SP app from its provisioned identity. Both sp-airline/server.ts and
// sp-hotel/server.ts are three-line wrappers around this.

import 'dotenv/config';
import { createRequire } from 'node:module';
import { dirname } from 'node:path';
import { buildStatusListCredential, createStatusList } from '@helixid/sdk-js';
import { env, type SpDefinition } from '../helixid-config/index.js';
import { createSpApp } from './app.js';
import { createAuditEmitter } from './audit.js';
import { loadSpIdentity, statePath, STATUS_LIST_LENGTH } from './identity.js';
import { SpStore } from './store.js';

const requireFromHere = createRequire(import.meta.url);

export async function serveSp(definition: SpDefinition): Promise<void> {
  const identity = await loadSpIdentity(env.walletsDir, definition.id);
  const store = await SpStore.open(
    statePath(env.walletsDir, definition.id),
    buildStatusListCredential('1', createStatusList(STATUS_LIST_LENGTH), identity.did, identity.baseUrl),
  );

  const { app } = createSpApp({
    definition,
    issuer: {
      did: identity.did,
      privateKeyHex: identity.privateKeyHex,
      publicKeyHex: identity.publicKeyHex,
    },
    baseUrl: identity.baseUrl,
    helixApiUrl: env.helixApiUrl,
    // Scope resolution happens inside this same container. The public did:web
    // base URL can be localhost, but the server should call its own loopback.
    mcpServerUrl: `http://127.0.0.1:${definition.port}/api/mcp`,
    store,
    // @helixid/widget is an ordinary dependency now, so let Node resolve it
    // rather than assuming a sibling checkout. Its "." export is dist/index.js,
    // so that file's directory is the dist the consent page loads /widget/*
    // from. Resolving also survives pnpm putting the package in a
    // content-addressed store path that no relative path could name.
    widgetDistPath: dirname(requireFromHere.resolve('@helixid/widget')),
    // Reports this SP's own verification and authorization decisions to the
    // shared audit log. Unconfigured (no URL/key) degrades to a no-op emitter.
    audit: createAuditEmitter({
      helixApiUrl: env.helixApiUrl,
      adminApiKey: env.adminApiKey,
      serviceDid: identity.did,
      serviceName: definition.displayName,
      onError: (message) => console.warn(`[${definition.id}] ${message}`),
    }),
  });

  app.listen(definition.port, '0.0.0.0', () => {
    console.log(`[${definition.id}] ${definition.displayName} listening on :${definition.port}`);
    console.log(`[${definition.id}] did:web  ${identity.did}`);
    console.log(`[${definition.id}] status   ${identity.statusListUrl}`);
    console.log(`[${definition.id}] consent  ${identity.baseUrl}/consent`);
  });
}
